From d46fd5a6e433b192f68448caad96d03706728279 Mon Sep 17 00:00:00 2001
From: rleix <rayx.lei@intel.com>
Date: Wed, 27 Jun 2018 09:42:03 +0800
Subject: [PATCH 1/2] [Gallery2] Add to handle the OnAudioFocusChangeListener

Fix the video playback playing all the time even the audiofocus is lost.

Change-Id: I402887a77651d53ae05e966cdde54edccb351a68
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-65418
Signed-off-by: Lei,RayX <rayx.lei@intel.com>
---
 src/com/android/gallery3d/app/MovieActivity.java | 20 +++++++++++++++-----
 src/com/android/gallery3d/app/MoviePlayer.java   |  8 +++++++-
 2 files changed, 22 insertions(+), 6 deletions(-)

diff --git a/src/com/android/gallery3d/app/MovieActivity.java b/src/com/android/gallery3d/app/MovieActivity.java
index 1547f6f..31cddcd 100644
--- a/src/com/android/gallery3d/app/MovieActivity.java
+++ b/src/com/android/gallery3d/app/MovieActivity.java
@@ -210,29 +210,39 @@ public class MovieActivity extends Activity {
         return false;
     }
 
+    private final AudioManager.OnAudioFocusChangeListener mListener =
+        new AudioManager.OnAudioFocusChangeListener() {
+            @Override
+            public void onAudioFocusChange(int focusChange) {
+                if (focusChange == AudioManager.AUDIOFOCUS_LOSS) {
+                    mPlayer.onPause();
+                }
+            }
+    };
+
     @Override
     public void onStart() {
-        ((AudioManager) getSystemService(AUDIO_SERVICE))
-                .requestAudioFocus(null, AudioManager.STREAM_MUSIC,
-                AudioManager.AUDIOFOCUS_GAIN_TRANSIENT);
         super.onStart();
     }
 
     @Override
     protected void onStop() {
-        ((AudioManager) getSystemService(AUDIO_SERVICE))
-                .abandonAudioFocus(null);
         super.onStop();
     }
 
     @Override
     public void onPause() {
+        ((AudioManager) getSystemService(AUDIO_SERVICE))
+                .abandonAudioFocus(mListener);
         mPlayer.onPause();
         super.onPause();
     }
 
     @Override
     public void onResume() {
+        ((AudioManager) getSystemService(AUDIO_SERVICE))
+                .requestAudioFocus(mListener, AudioManager.STREAM_MUSIC,
+                AudioManager.AUDIOFOCUS_GAIN);
         mPlayer.onResume();
         super.onResume();
     }
diff --git a/src/com/android/gallery3d/app/MoviePlayer.java b/src/com/android/gallery3d/app/MoviePlayer.java
index f6bd367..48fa535 100644
--- a/src/com/android/gallery3d/app/MoviePlayer.java
+++ b/src/com/android/gallery3d/app/MoviePlayer.java
@@ -94,6 +94,7 @@ public class MoviePlayer implements
 
     // If the time bar is visible.
     private boolean mShowing;
+    private boolean mPrepared = false;
 
     private Virtualizer mVirtualizer;
 
@@ -121,6 +122,7 @@ public class MoviePlayer implements
         mContext = movieActivity.getApplicationContext();
         mRootView = rootView;
         mVideoView = (VideoView) rootView.findViewById(R.id.surface_view);
+        mVideoView.setAudioFocusRequest(AudioManager.AUDIOFOCUS_NONE);
         mBookmarker = new Bookmarker(movieActivity);
         mUri = videoUri;
 
@@ -154,6 +156,7 @@ public class MoviePlayer implements
         mVideoView.setOnPreparedListener(new MediaPlayer.OnPreparedListener() {
             @Override
             public void onPrepared(MediaPlayer player) {
+                mPrepared = true;
                 if (!mVideoView.canSeekForward() || !mVideoView.canSeekBackward()) {
                     mController.setSeekable(false);
                 } else {
@@ -277,9 +280,12 @@ public class MoviePlayer implements
     public void onPause() {
         mHasPaused = true;
         mHandler.removeCallbacksAndMessages(null);
-        mVideoPosition = mVideoView.getCurrentPosition();
+        if (mPrepared) {
+            mVideoPosition = mVideoView.getCurrentPosition();
+        }
         mBookmarker.setBookmark(mUri, mVideoPosition, mVideoView.getDuration());
         mVideoView.suspend();
+        mPrepared = false;
         mResumeableTime = System.currentTimeMillis() + RESUMEABLE_TIMEOUT;
     }
 
-- 
1.9.1

